src/metabolic_ninja/worker/designer.py

```diff
@@ -0,0 +1,285 @@
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#    http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+import logging
+
+import cameo.core.target
+from cameo.strain_design import DifferentialFVA, OptGene
+from cameo.strain_design.heuristic.evolutionary.objective_functions import (
+    biomass_product_coupled_min_yield,
+    product_yield,
+)
+from cameo.strain_design.heuristic.evolutionary_based import CofactorSwapOptimization
+from cobra.exceptions import OptimizationError
+from numpy import isnan
+
+from .evaluate import evaluate_biomass_coupled_production, evaluate_production
+from .helpers import find_synthetic_reactions, manipulation_helper
+
+
+logger = logging.getLogger(__name__)
+
+
+def differential_fva_optimization(pathway, model):
+    """
+    Compare FVA results on the production plane with maximum growth.
+
+    Parameters
+    ----------
+    pathway :
+        A heterologous pathway identified by cameo.
+    model : cobra.Model
+        The model under investigation.
+
+    Returns
+    -------
+    cameo.design
+        A number of differential FVA designs corresponding to evenly spaced
+        points on the surface of the phenotypic phase plane.
+
+    """
+    with model:
+        pathway.apply(model)
+        predictor = DifferentialFVA(
+            design_space_model=model,
+            objective=pathway.product.id,
+            variables=[model.biomass],
+            normalize_ranges_by=model.biomass,
+            # Excluding the maxima, this corresponds to three evenly spaced
+            # designs.
+            points=5,
+        )
+        try:
+            designs = predictor.run(progress=False)
+        except ZeroDivisionError as error:
+            logger.warning(
+                "Encountered the following error in DiffFVA.", exc_info=error
+            )
+            designs = None
+    return designs
+
+
+def evaluate_diff_fva(designs, pathway, model, method):
+    """Evaluate the differential FVA designs."""
+    if designs is None:
+        return []
+    logger.info(
+        f"Evaluating {len(designs) - 1} differential FVA surface points."
+    )
+    results = []
+    designs = list(designs)
+    with model:
+        pathway.apply(model)
+        # The reference point is automatically ignored. Thus four of the
+        # original five points remain. The first point in order represents
+        # maximum production and zero growth. We ignore the point of lowest
+        # production (the last one in order).
+        for design_result in designs[:-1]:
+            with model:
+                design_result.apply(model)
```

Comment:
In certain cases applying the design may lead to an infeasible model. Thus `evaluate_production` and `evaluate_biomass_coupled_production` below my raise an `OptimizationError`.

Reply:
That's also the case for co-factor swap and OptGene.

Reply:
Thanks for spotting this, but I think it's out of the scope of this PR as the logic hasn't been modified. Could you raise it in another channel, for example by creating a new issue, please?